### DISCUSSION TITLE
aes和sm4s盒复合域实现方法

### DISCUSSION THREAD

**Discussion by creed-sdu** [emmansun/gmsm#5] [General]
我的博客地址在这:) 
http://zongyue.top:8090/archives/aes%E5%92%8Csm4s%E7%9B%92%E5%A4%8D%E5%90%88%E5%9F%9F%E5%AE%9E%E7%8E%B0%E6%96%B9%E6%B3%95

**Comment by emmansun** [comment]
这个版本的编排好多了，感谢！

**Comment by emmansun** [comment]
Hi @creed-sd 对于“My question is how to generate all (M1, C1, M2, C2)?”，你有解吗？

**Reply by creed-sdu** [reply]
看上去可以选不同的单位根，得到另一组解，我没有试过 :)

**Reply by emmansun** [reply]
[sm4ni](https://github.com/mjosaarinen/sm4ni)里用到的这组：
```
M1= 0x14, 0x2e, 0x16, 0x8a, 0x60, 0x0d, 0x9b, 0x66
C1= 0x01
M2= 0xfe, 0x54, 0xaf, 0xdd, 0xf7, 0xf9, 0xac, 0xe2
C2= 0x34
```
不知道怎么算出来的。

**Comment by emmansun** [comment]
经过测试，在AES/SM4采用同样的算法及参数的情况下，无论[使用多项式基](https://gist.github.com/emmansun/fa3b11833fc11d6832f0aa14553caf47)还是[使用正规基](https://gist.github.com/emmansun/9f8d449072f6048538add79bd146b661)，无论GF(2^4) Nu = Cz + D，C/D采用论文[A very compact Rijndael S-box](https://calhoun.nps.edu/bitstream/handle/10945/791/NPS-MA-04-001.pdf?sequence=1&isAllowed=y)中的何种组合，得到的{M1, C1, M2, C2}结果一样。

**Reply by creed-sdu** [reply]
那可能还是得考虑别的思路，我当时主要是为了推塔域实现，然后顺带想到的可以这样推导下，看起来这个不是正路

**Reply by emmansun** [reply]
你的推导没有问题，不能说不是正路。我给[mjosaarinen](https://github.com/mjosaarinen)发了邮件，看看能否得到回复。

**Reply by emmansun** [reply]
其实主要是M1和M2的关系，从复合域推导出来的，M1和M2存在如下关系：
```
M2 = Asm4*(Asm4*M1.Inverse())*Aaes.Inverse()
```
但mjosaarionen的那一组不存在上述关系，或许存在其他未知确定性关系。C1和M1, C2和M2的关系是确定的，一致的：
```
C1 = M1 * Asm4.Inverse() * 0xd3
C2 = M2 * 0x63 + 0xd3
```

**Reply by emmansun** [reply]
没能收到[mjosaarinen](https://github.com/mjosaarinen)的回复。

**Comment by emmansun** [comment]
感谢AI！
<img width="907" height="634" alt="image" src="https://github.com/user-attachments/assets/33b5af3a-7b03-481e-a537-bd535a0b2f51" />

[python脚本](https://gist.github.com/emmansun/0161916f775dd5230f99b7b573ec96b7)
